libsnark/zk_proof_systems/plonk/prover.tcc

```diff
@@ -0,0 +1,1186 @@
+    // TODO: move to unit test for
+    // plonk_interpolate_polynomial_from_points
+#ifdef DEBUG_PLONK
+    for (int i = 0; i < nwitness; ++i) {
+        printf("[%s:%d] this->W_polys[%d]\n", __FILE__, __LINE__, i);
+        print_vector(W_polys[i]);
+        assert(W_polys[i] == example.W_polys[i]);
+    }
+#endif // #ifdef DEBUG_PLONK
+
+    // hard-coded values for the "random" blinding constants from
+    // example circuit
+    blind_scalars = example.prover_blind_scalars;
+    // represent the blinding scalars b1, b2, ..., b9 as polynomials
+    std::vector<std::vector<Field>> blind_polys{
+        {blind_scalars[1], blind_scalars[0]}, // b1 + b0 X
+        {blind_scalars[3], blind_scalars[2]}, // b3 + b2 X
+        {blind_scalars[5], blind_scalars[4]}  // b5 + b4 X
+    };
+
+    // compute blinded witness polynomials e.g. a_poly =
+    // blind_polys[0] * zh_poly + W_polys[0]
+    W_polys_blinded.resize(nwitness);
+    for (int i = 0; i < nwitness; ++i) {
+        libfqfft::_polynomial_multiplication<Field>(
+            W_polys_blinded[i], blind_polys[i], round_zero_out.zh_poly);
+        libfqfft::_polynomial_addition<Field>(
+            W_polys_blinded[i], W_polys_blinded[i], W_polys[i]);
+    }
+    // evaluate blinded witness polynomials at the secret input
+    W_polys_blinded_at_secret_g1.resize(W_polys_blinded.size());
+    plonk_evaluate_polys_at_secret_G1<ppT>(
+        srs.secret_powers_g1, W_polys_blinded, W_polys_blinded_at_secret_g1);
+
+    round_one_out_t<ppT> round_one_out(
+        std::move(blind_scalars),
+        std::move(W_polys),
+        std::move(W_polys_blinded),
+        std::move(W_polys_blinded_at_secret_g1));
+
+    return round_one_out;
+}
+
+/// Prover Round 2
+///
+/// INPUT
+/// \param[in] blind_scalars: blinding scalars b1, b2, ..., b9 (only
+///            b7,b8,b9 used in round 2) (from round 1)
+/// \param[in] zh_poly: vanishing polynomial Zh (from round 0)
+/// \param[in] witness: witness values
+/// \param[in] srs: structured reference string containing also
+///            circuit-specific information
+///
+/// OUTPUT
+/// \param[out] beta, gamma: permutation challenges -- hashes of
+/// transcript
+/// \param[out] z_poly: blinded accumulator poly z(x)
+/// \param[out] z_poly_at_secret_g1: blinded accumulator poly z(x)
+///             evaluated at secret
+template<typename ppT>
+round_two_out_t<ppT> plonk_prover<ppT>::round_two(
+    const round_zero_out_t<ppT> &round_zero_out,
+    const round_one_out_t<ppT> &round_one_out,
+    const std::vector<libff::Fr<ppT>> &witness,
+    const srs<ppT> &srs)
+{
+    using Field = libff::Fr<ppT>;
+    // initialize hard-coded values from example circuit
+    plonk_example example;
+
+    // output from round 2
+    libff::Fr<ppT> beta;
+    libff::Fr<ppT> gamma;
+    polynomial<libff::Fr<ppT>> z_poly;
+    libff::G1<ppT> z_poly_at_secret_g1;
+
+    // permutation challenges (hashes of transcript); fixed to test
+    // values
+    beta = example.beta;
```

Comment:
I think these can be passed in by the caller, who can compute them from the transcript.  The example values can be passed in for the purposes of the unit test.

Reply:
Very good suggestion. Addressed.